// @polywrap/client-js/build/wasm/WasmWrapper.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WasmWrapper = void 0;
var imports_1 = require("./imports");
var core_js_1 = require("@polywrap/core-js");
var wrap_manifest_types_js_1 = require("@polywrap/wrap-manifest-types-js");
var msgpack_js_1 = require("@polywrap/msgpack-js");
var tracing_js_1 = require("@polywrap/tracing-js");
var asyncify_js_1 = require("@polywrap/asyncify-js");
var WasmWrapper = /** @class */ (function (_super) {
    __extends(WasmWrapper, _super);
    function WasmWrapper(_uri, _manifest, _uriResolver, _clientEnv) {
        var _this = _super.call(this) || this;
        _this._uri = _uri;
        _this._manifest = _manifest;
        _this._uriResolver = _uriResolver;
        _this._clientEnv = _clientEnv;
        _this._info = undefined;
        _this._wasm = undefined;
        tracing_js_1.Tracer.startSpan("WasmWrapper: constructor");
        tracing_js_1.Tracer.setAttribute("args", {
            uri: _this._uri,
            manifest: _this._manifest,
            clientEnv: _this._clientEnv,
            uriResolver: _this._uriResolver,
        });
        tracing_js_1.Tracer.endSpan();
        return _this;
    }
    WasmWrapper.prototype.getFile = function (options, client) {
        return __awaiter(this, void 0, void 0, function () {
            var path, encoding, _a, data, error, decoder, text;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        path = options.path, encoding = options.encoding;
                        return [4 /*yield*/, core_js_1.UriResolverInterface.module.getFile({
                                invoke: function (options) { return client.invoke(options); },
                            }, 
                            // TODO: support all types of URI resolvers (cache, etc)
                            new core_js_1.Uri(this._uriResolver), core_js_1.combinePaths(this._uri.path, path))];
                    case 1:
                        _a = _b.sent(), data = _a.data, error = _a.error;
                        if (error) {
                            throw error;
                        }
                        // If nothing is returned, the file was not found
                        if (!data) {
                            throw Error("WasmWrapper: File was not found.\nURI: " + this._uri + "\nSubpath: " + path);
                        }
                        if (encoding) {
                            decoder = new TextDecoder(encoding);
                            text = decoder.decode(data);
                            if (!text) {
                                throw Error("WasmWrapper: Decoding the file's bytes array failed.\nBytes: " + data);
                            }
                            return [2 /*return*/, text];
                        }
                        return [2 /*return*/, data];
                }
            });
        });
    };
    WasmWrapper.prototype.getManifest = function (options, client) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, moduleManifest, data;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this._info !== undefined) {
                            return [2 /*return*/, this._info];
                        }
                        _a = this;
                        return [4 /*yield*/, this.getFile({
                                path: "schema.graphql",
                                encoding: "utf-8",
                            }, client)];
                    case 1:
                        _a._schema = (_b.sent());
                        moduleManifest = "wrap.info";
                        return [4 /*yield*/, this.getFile({ path: moduleManifest }, client)];
                    case 2:
                        data = (_b.sent());
                        if (!data) {
                            throw Error("Package manifest does not contain information");
                        }
                        return [2 /*return*/, wrap_manifest_types_js_1.deserializeWrapManifest(data)];
                }
            });
        });
    };
    WasmWrapper.prototype.invoke = function (options, client) {
        return __awaiter(this, void 0, void 0, function () {
            var method_1, args_1, wasm, state, abort, memory, instance, exports_1, result, invokeResult, error_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        method_1 = options.method;
                        args_1 = options.args || {};
                        return [4 /*yield*/, this._getWasmModule(client)];
                    case 1:
                        wasm = _a.sent();
                        state = {
                            invoke: {},
                            subinvoke: {
                                args: [],
                            },
                            subinvokeImplementation: {
                                args: [],
                            },
                            invokeResult: {},
                            method: method_1,
                            args: core_js_1.isBuffer(args_1) ? args_1 : msgpack_js_1.msgpackEncode(args_1),
                            env: msgpack_js_1.msgpackEncode(this._getClientEnv()),
                        };
                        abort = function (message) {
                            throw new Error("WasmWrapper: Wasm module aborted execution.\nURI: " + _this._uri.uri + "\n" +
                                ("Method: " + method_1 + "\n") +
                                ("Args: " + JSON.stringify(args_1, null, 2) + "\nMessage: " + message + ".\n"));
                        };
                        memory = asyncify_js_1.AsyncWasmInstance.createMemory({ module: wasm });
                        return [4 /*yield*/, asyncify_js_1.AsyncWasmInstance.createInstance({
                                module: wasm,
                                imports: imports_1.createImports({
                                    state: state,
                                    client: client,
                                    memory: memory,
                                    abort: abort,
                                }),
                                requiredExports: WasmWrapper.requiredExports,
                            })];
                    case 2:
                        instance = _a.sent();
                        exports_1 = instance.exports;
                        return [4 /*yield*/, exports_1._wrap_invoke(state.method.length, state.args.byteLength, state.env.byteLength)];
                    case 3:
                        result = _a.sent();
                        invokeResult = this._processInvokeResult(state, result, abort);
                        switch (invokeResult.type) {
                            case "InvokeError": {
                                throw Error("WasmWrapper: invocation exception encountered.\n" +
                                    ("uri: " + this._uri.uri + "\n") +
                                    ("method: " + method_1 + "\n") +
                                    ("args: " + JSON.stringify(args_1, null, 2) + "\n") +
                                    ("exception: " + invokeResult.invokeError));
                            }
                            case "InvokeResult": {
                                return [2 /*return*/, {
                                        data: invokeResult.invokeResult,
                                        encoded: true,
                                    }];
                            }
                            default: {
                                throw Error("WasmWrapper: Unknown state \"" + state + "\"");
                            }
                        }
                        return [3 /*break*/, 5];
                    case 4:
                        error_1 = _a.sent();
                        return [2 /*return*/, {
                                error: error_1,
                            }];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    WasmWrapper.prototype.getSchema = function (client) {
        return __awaiter(this, void 0, void 0, function () {
            var schema, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this._schema) {
                            return [2 /*return*/, this._schema];
                        }
                        schema = "schema.graphql";
                        _a = this;
                        return [4 /*yield*/, this.getFile({ path: schema, encoding: "utf8" }, client)];
                    case 1:
                        _a._schema = (_b.sent());
                        return [2 /*return*/, this._schema];
                }
            });
        });
    };
    WasmWrapper.prototype._processInvokeResult = function (state, result, abort) {
        if (result) {
            if (!state.invoke.result) {
                abort("Invoke result is missing.");
            }
            return {
                type: "InvokeResult",
                invokeResult: state.invoke.result,
            };
        }
        else {
            if (!state.invoke.error) {
                abort("Invoke error is missing.");
            }
            return {
                type: "InvokeError",
                invokeError: state.invoke.error,
            };
        }
    };
    WasmWrapper.prototype._getClientEnv = function () {
        var _a;
        if (!((_a = this._clientEnv) === null || _a === void 0 ? void 0 : _a.env)) {
            return {};
        }
        return this._clientEnv.env;
    };
    WasmWrapper.prototype._getWasmModule = function (client) {
        return __awaiter(this, void 0, void 0, function () {
            var moduleManifest, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this._wasm !== undefined) {
                            return [2 /*return*/, this._wasm];
                        }
                        moduleManifest = "wrap.wasm";
                        if (!moduleManifest) {
                            throw Error("Package manifest does not contain a definition for module");
                        }
                        return [4 /*yield*/, this.getFile({ path: moduleManifest }, client)];
                    case 1:
                        data = (_a.sent());
                        this._wasm = data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    WasmWrapper.requiredExports = ["_wrap_invoke"];
    __decorate([
        tracing_js_1.Tracer.traceMethod("WasmWrapper: getFile")
    ], WasmWrapper.prototype, "getFile", null);
    __decorate([
        tracing_js_1.Tracer.traceMethod("WasmWrapper: getManifest")
    ], WasmWrapper.prototype, "getManifest", null);
    __decorate([
        tracing_js_1.Tracer.traceMethod("WasmWrapper: invoke")
    ], WasmWrapper.prototype, "invoke", null);
    __decorate([
        tracing_js_1.Tracer.traceMethod("WasmWrapper: getSchema")
    ], WasmWrapper.prototype, "getSchema", null);
    __decorate([
        tracing_js_1.Tracer.traceMethod("WasmWrapper: _processInvokeResult")
    ], WasmWrapper.prototype, "_processInvokeResult", null);
    __decorate([
        tracing_js_1.Tracer.traceMethod("WasmWrapper: _getClientEnv")
    ], WasmWrapper.prototype, "_getClientEnv", null);
    __decorate([
        tracing_js_1.Tracer.traceMethod("WasmWrapper: getWasmModule")
    ], WasmWrapper.prototype, "_getWasmModule", null);
    return WasmWrapper;
}(core_js_1.Wrapper));
exports.WasmWrapper = WasmWrapper;
