// @polywrap/ipfs-resolver-plugin-js/build/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.ipfsResolverPlugin = exports.IpfsResolverPlugin = void 0;
var wrap_1 = require("./wrap");
// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
var isIPFS = require("is-ipfs");
var IpfsResolverPlugin = /** @class */ (function (_super) {
    __extends(IpfsResolverPlugin, _super);
    function IpfsResolverPlugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // uri-resolver.core.web3api.eth
    IpfsResolverPlugin.prototype.tryResolveUri = function (args, _client) {
        return __awaiter(this, void 0, void 0, function () {
            var manifestSearchPattern, manifest, manifestResult, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (args.authority !== "ipfs") {
                            return [2 /*return*/, null];
                        }
                        if (!IpfsResolverPlugin.isCID(args.path)) {
                            // Not a valid CID
                            return [2 /*return*/, { manifest: null, uri: null }];
                        }
                        manifestSearchPattern = "wrap.info";
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, wrap_1.Ipfs_Module.cat({
                                cid: args.path + "/" + manifestSearchPattern,
                                options: {
                                    timeout: 5000,
                                },
                            }, _client)];
                    case 2:
                        manifestResult = _a.sent();
                        if (manifestResult.data) {
                            manifest = Buffer.from(manifestResult.data);
                        }
                        else {
                            throw new Error();
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, { uri: null, manifest: manifest !== null && manifest !== void 0 ? manifest : null }];
                }
            });
        });
    };
    IpfsResolverPlugin.prototype.getFile = function (args, client) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var resolveResult, result, catResult, e_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, wrap_1.Ipfs_Module.resolve({
                                cid: args.path,
                                options: {
                                    timeout: 5000,
                                },
                            }, client)];
                    case 1:
                        resolveResult = _b.sent();
                        result = resolveResult.data;
                        if (!result) {
                            return [2 /*return*/, null];
                        }
                        return [4 /*yield*/, wrap_1.Ipfs_Module.cat({
                                cid: result.cid,
                                options: {
                                    provider: result.provider,
                                    timeout: 20000,
                                    disableParallelRequests: true,
                                },
                            }, client)];
                    case 2:
                        catResult = _b.sent();
                        return [2 /*return*/, (_a = catResult.data) !== null && _a !== void 0 ? _a : null];
                    case 3:
                        e_2 = _b.sent();
                        return [2 /*return*/, null];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    IpfsResolverPlugin.isCID = function (cid) {
        return isIPFS.cid(cid) || isIPFS.cidPath(cid) || isIPFS.ipfsPath(cid);
    };
    return IpfsResolverPlugin;
}(wrap_1.Module));
exports.IpfsResolverPlugin = IpfsResolverPlugin;
exports.ipfsResolverPlugin = function (opts) {
    return {
        factory: function () { return new IpfsResolverPlugin(opts); },
        manifest: wrap_1.manifest,
    };
};
exports.plugin = exports.ipfsResolverPlugin;
